docs/firmware-architecture-and-design.md

```diff
@@ -0,0 +1,119 @@
+ */
+Dog_t* dog_create(DogColor color, float (*attempt_to_run_at_speed)(float)); 
+
+/**
+ * Destroy the given Dog, freeing any memory allocated for it
+ * @param dog The Dog to destroy
+ */
+void dog_destroy(Dog_t* dog);
+
+/**
+ * Get the color of a dog
+ * @param dog The dog to get the color of
+ * @return The color of the given dog
+ */
+DogColor dog_getColor(Dog_t* dog);
+
+/**
+ * Attempt to set the speed of a Dog
+ * @param desired_speed The speed to attempt, in m/s
+ * @return The maximum speed actually achieved, in m/s
+ */
+float dog_attemptToRunAtSpeed(Dog_t* dog, float desired_speed);
+```
+
+There are a few things to note here: 
+* We have declared a `struct Dog` and typedef'd it to just `Dog_t`, but we *haven't actually said what's in the struct*. This is a crucial point, because it means that anything that includes `dog.h` can only interact with a `struct Dog` through the functions we define here. 
+* We passed a function into `dog_create`. This is a really useful technique that is part of what enables us to do stuff like providing different "implementations" of a robot to the `app` layer depending on if we're in simulation or running on the real robots.
+
+We would implement this "class" as follows:
+
+_dog.c_
+``` C
+#include "dog.h"
+
+struct Dog {
+    DogColor color;
+    float (*attempt_to_run_at_speed)(float);
+};
+
+Dog_t* dog_create(DogColor color, float (*attempt_to_run_at_speed)(float)){
+    // NOTE: We can't do this outside this `.c` file, because the only place
+    //       that we know the size of the `Dog` struct is here.
+    Dog_t* dog = (Dog_t*)(malloc(sizeof(Dog_t)));
+
+    dog->color = color;
+    dog->attempt_to_run_at_speed = attempt_to_run_at_speed;
+
+    return dog;
+}
+
+void dog_destroy(Dog_t* dog){
+    // NOTE: We can't do this outside this `.c` file, because the only place
+    //       that we know the size of the `Dog` struct is here.
+    free(dog);
+}
+
+DogColor dog_getColor(Dog_t* dog){
+    return dog->color;
+}
+
+float dog_attemptToRunAtSpeed(Dog_t* dog, float desired_speed){
+    return dog->attempt_to_run_at_speed(desired_speed);
+}
+```
+
+and we would use this class as follows:
+``` C
+#include "dog.h"
+
+// Assume we get our "attempt_to_run_at_speed" method from this header
+#include "dog_simulator.h"
+
+Dog_t* my_dog = dog_create(BROWN, &simulated_dog_attempt_speed);
+
+assert(dog_getColor(my_dog) == BROWN);
+
+float achieved_speed = dog_attemptToRunAtSpeed(my_dog, 9001);
+
+dog_destroy(my_dog);
+```
+
+# App/IO Split
+The firmware is divided into `app` and `io` layers. 
+* The `app` layer contains anything that is portable between all our robots and our simulator. 
+* The `io` layer contains everything below this. 
+* The interface between these layers is the `FirmwareWorld` class, which represents an abstraction of the world from the perspective of the robot. **This is the only interface that the `app` layer should ever be able to use to access the outside world**.
```

Comment:
I feel like the explanation for the split is a bit short here, given how important it is. Would really appreciate input from someone who hasn't really seen the firmware here to see if this is sufficient given no context.

Reply:
This definitely felt very short when I wrote it, but I couldn't think of anything else to add that would significantly help explain things :confused:

Reply:
I think it's sufficient, but a bit more detail on what each layer does or examples of functionalities in each layer might be helpful

Reply:
Added examples to `app` and `io`, let me know what you think

Reply:
The examples definitely help with explaining the layers, looks good to me 👍